llrt_core/src/vm.rs

```diff
@@ -700,3 +713,58 @@ fn set_import_meta(module: &Module<'_>, filepath: &str) -> Result<()> {
     meta.prop("url", ["file://", filepath].concat())?;
     Ok(())
 }
+
+fn get_module_path_in_node_modules(ctx: &Ctx<'_>, specifier: &str) -> Result<String> {
+    let current_dir = env::current_dir()
+        .ok()
+        .and_then(|path| path.into_os_string().into_string().ok())
+        .unwrap_or_else(|| "/".to_string());
+
+    let (scope, name) = match specifier.split_once('/') {
+        Some((s, n)) => (s, ["./", n].concat()),
+        None => (specifier, ".".to_string()),
+    };
+
+    let mut package_json_path = [&current_dir, "/node_modules/"].concat();
+    let base_path_length = package_json_path.len();
+    package_json_path.push_str(scope);
+    package_json_path.push_str("/package.json");
+
+    let (scope, name) = if name != "." && fs::metadata(&package_json_path).is_err() {
+        package_json_path.truncate(base_path_length);
+        package_json_path.push_str(specifier);
+        package_json_path.push_str("/package.json");
+        (specifier, ".")
+    } else {
+        (scope, name.as_str())
+    };
+
+    if fs::metadata(&package_json_path).is_err() {
+        return Err(Exception::throw_reference(
+            ctx,
+            &["Error resolving module '", specifier, "'"].concat(),
+        ));
+    };
+
+    let package_json = fs::read(&package_json_path).unwrap_or_default();
+    let package_json = simd_json::to_owned_value(&mut package_json.clone()).or_throw(ctx)?;
+
+    let module_path = json_as_str(&package_json, &["exports", name, "require", "default"])
+        .or_else(|| json_as_str(&package_json, &["exports", name, "require"]))
+        .or_else(|| json_as_str(&package_json, &["exports", "require", "default"]))
+        .or_else(|| json_as_str(&package_json, &["exports", "default"]))
+        .or_else(|| json_as_str(&package_json, &["main"]))
+        .unwrap_or("./index.js");
```

Comment:
I wonder here if we can do this a bit more elegantly. We sacrifice a bit of readability for better performance as we don't check each key several times and iterate trough the object multiple times:
```rust
let package_json = simd_json::to_borrowed_value(&mut json).or_throw(ctx)?;
let module_path = get_module_path(&package_json, name)?;

...

fn get_module_path<'a>(package_json: &'a BorrowedValue<'a>, name: &str) -> Result<&'a str> {
    if let BorrowedValue::Object(map) = package_json {
        if let Some(BorrowedValue::Object(exports)) = map.get("exports") {
            // Check for nested structure: exports -> name
            if let Some(BorrowedValue::Object(name_obj)) = exports.get(name) {
                // Check for name -> require -> default
                if let Some(BorrowedValue::Object(require)) = name_obj.get("require") {
                    if let Some(BorrowedValue::String(default)) = require.get("default") {
                        return Ok(default.as_ref());
                    }
                }
                // Check for name -> default
                if let Some(BorrowedValue::String(default)) = name_obj.get("default") {
                    return Ok(default.as_ref());
                }
            }

            // Check for nested structure: exports -> require -> default
            if let Some(BorrowedValue::Object(require)) = exports.get("require") {
                if let Some(BorrowedValue::String(default)) = require.get("default") {
                    return Ok(default.as_ref());
                }
            }

            // Check for exports -> default
            if let Some(BorrowedValue::String(default)) = exports.get("default") {
                return Ok(default.as_ref());
            }
        }

        // Check for main field
        if let Some(BorrowedValue::String(main)) = map.get("main") {
            return Ok(main.as_ref());
        }
    }

    Ok("./index.js")
}
```

Also no need to clone the Vec we read from the package ison path, simply return a mutable reference to it.

Reply:
Certainly performance should have been a top priority here. And this is still easy enough to see!
We have adopted it as is, with one missing search and a slight modification to the comment position.